File: app/src/main/java/com/ma/lightweather/widget/HourWeatherView.java
package com.ma.lightweather.widget;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.CornerPathEffect;
import android.graphics.Paint;
import android.graphics.Path;
import android.util.AttributeSet;
import android.util.Log;
import android.view.View;

import com.ma.lightweather.R;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * Created by Aeolus on 2018/8/24.
 */

public class HourWeatherView extends View {

    private Paint pointPaint=new Paint();
    private Paint outPointPaint=new Paint();
    private Paint textPaint=new Paint();
    private Paint tmpPaint =new Paint();
    private Path path=new Path();

    private List<Integer> tmpList=new ArrayList<>();
    private List<Float> midelTmpList=new ArrayList<>();
    private List<Integer> popList=new ArrayList<>();
    private List<String> dateList=new ArrayList<>();
    private List<String> txtList=new ArrayList<>();
    private List<String> dirList=new ArrayList<>();

    private static int lineWidth=4;
    private static int pointWidth=3;
    private static int outPointWidth=3;
    private static int pointRadius=5;
    private static int outPointRadius=10;

    private int xSpace;
    private int offsetHigh;
    private int ySpace;
    private int max;
    private int min;
    private int textHigh;
    private int textSpace;

    public HourWeatherView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    private void init() {

        pointPaint.setAntiAlias(true);
        pointPaint.setStrokeWidth(pointWidth);
        pointPaint.setStyle(Paint.Style.FILL);

        outPointPaint.setAntiAlias(true);
        outPointPaint.setStrokeWidth(outPointWidth);
        outPointPaint.setStyle(Paint.Style.STROKE);

        tmpPaint.setColor(getResources().getColor(R.color.temp));
        tmpPaint.setAntiAlias(true);
        tmpPaint.setStrokeWidth(lineWidth);
        tmpPaint.setStyle(Paint.Style.STROKE);

        textPaint.setColor(getResources().getColor(R.color.text));
        textPaint.setAntiAlias(true);
        textPaint.setTextAlign(Paint.Align.CENTER);

    }

    @Override
    protected void onDraw(Canvas canvas) {
        int viewhigh=getMeasuredHeight();
        int viewwidth=getMeasuredWidth();
        textPaint.setTextSize(viewwidth/30);
        Paint.FontMetrics fm = textPaint.getFontMetrics();
        textHigh =(int)Math.ceil(fm.bottom - fm.top);
        textSpace=(int)Math.ceil(fm.bottom - fm.leading)+(int)Math.ceil(fm.ascent - fm.top);
        offsetHigh=3*textHigh;
        xSpace=viewwidth/16;
        ySpace=(viewhigh-2*offsetHigh)/50;
        if(tmpList.size()>0) {
            max = Collections.max(tmpList);
            min = Collections.min(tmpList);
            ySpace = (viewhigh-2*offsetHigh) / (max - min);
        }
        //getFoldLine(canvas);
        getCurveLine(canvas);
        //日期
        for (int i=0;i<dateList.size();i++){
            String[] data1=dateList.get(i).split(" ");
            canvas.drawText(data1[1],getX(2*i+1), textHigh,textPaint);
        }
        //天气状况
        for (int i=0;i<txtList.size();i++){
            canvas.drawText(txtList.get(i),getX(2*i+1),viewhigh-textSpace,textPaint);
        }
        //降水概率
        for (int i=0;i<popList.size();i++){
            canvas.drawText(popList.get(i)+"%",getX(2*i+1),viewhigh-textSpace- textHigh,textPaint);
        }
        //分割线
        for (int i=1;i<dateList.size();i++){
            canvas.drawLine(getX(2*i),50,getX(2*i),viewhigh- textHigh,textPaint);
        }
    }

    private void  getFoldLine(Canvas canvas){
        for(int i=0;i<tmpList.size();i++) {
            pointPaint.setColor(getResources().getColor(R.color.temp));
            outPointPaint.setColor(getResources().getColor(R.color.temp));
            pointPaint.setColor(getResources().getColor(R.color.temp));
            outPointPaint.setColor(getResources().getColor(R.color.temp));
            if (i < tmpList.size() - 1) {
                canvas.drawLine((2 * i + 1) * xSpace, (max - tmpList.get(i)) * ySpace + offsetHigh,
                        (2 * i + 3) * xSpace, (max - tmpList.get(i + 1)) * ySpace + offsetHigh, tmpPaint);
            }

//            canvas.drawCircle(getX(2 * i + 1), getY(i, tmpList), pointRadius, pointPaint);
//            canvas.drawCircle(getX(2 * i + 1), getY(i, tmpList), outPointRadius, outPointPaint);
        }
    }

    /**
     * 需要四个点控制   当前点  当前点后控制点  下一个点  下一个点前控制点
     *
     */
    private void  getCurveLine(Canvas canvas){
        path.reset();
        for(int i=0;i<tmpList.size();i++){
            float k1=0;//当前点斜率
            float k2=0;//下一个点斜率
            float b1=0;//当前点常量
            float b2=0;//下一个点常量
            float x1=0;// 当前点后控制点x坐标
            float y1=0;// 当前点后控制点y坐标
            float x2=0;// 下一个点前控制点x坐标
            float y2=0;// 下一个点前控制点y坐标
            double r=0.5;//控制点与经过点的距离  取值0-1  越大距离越近
            if(i==0){
                x1=getX(2*i+2-r);
                y1=getY(i);
                k2=(getY(i+2)-getY(i))/getX(4);
                b2=getY(i+1)-k2*getX(2*i+3);
                x2=getX(2*i+2+r);
                y2=x2*k2+b2;
                path.moveTo(getX(2*i+1),getY(i));
                path.cubicTo(x1,y1, x2,y2, getX(2*i+3),getY(i+1));
            } else if(i<tmpList.size()-2){
                k1=(getY(i+1)-getY(i-1))/getX(4);
                b1=getY(i)-k1*getX(2*i+1);
                x1=getX(2*i+2-r);
                y1=x1*k1+b1;
                k2=(getY(i+2)-getY(i))/getX(4);
                b2=getY(i+1)-k2*getX(2*i+3);
                x2=getX(2*i+2+r);
                y2=x2*k2+b2;
                path.moveTo(getX(2*i+1),getY(i));
                path.cubicTo(x1,y1, x2,y2, getX(2*i+3),getY(i+1));
            }else if(i==tmpList.size()-2){
                k1=(getY(i+1)-getY(i-1))/getX(4);
                b1=getY(i)-k1*getX(2*i+1);
                x1=getX(2*i+2-r);
                y1=x1*k1+b1;
                x2=getX(2*i+2+r);
                y2=getY(i);
                path.moveTo(getX(2*i+1),getY(i));
                path.cubicTo(x1,y1, x2,y2, getX(2*i+3),getY(i+1));
            }
            canvas.drawPath(path,tmpPaint);
            canvas.drawText(tmpList.get(i)+"°",getX(2*i+1),getY(i)-textSpace,textPaint);
        }
    }

    public void loadViewData(List<Integer> tmpList, List<Integer> popList, List<String> dateList , List<String> txtList, List<String> dirList) {
        this.tmpList=tmpList;
        this.popList=popList;
        this.dateList=dateList;
        this.txtList=txtList;
        this.dirList=dirList;
        postInvalidate();
    }

    private float getX(double i){
        return Float.valueOf(String.valueOf(i*xSpace));
    }

    private float getY(int i){
        return (max-tmpList.get(i))*ySpace+offsetHigh;
    }

}

File: app/src/main/java/com/ma/lightweather/widget/HourWeatherView.kt
package com.ma.lightweather.widget

import android.animation.ValueAnimator
import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Path
import android.graphics.PathMeasure
import android.support.v4.content.ContextCompat
import android.util.AttributeSet
import android.view.View
import com.ma.lightweather.R
import com.ma.lightweather.model.Weather
import java.util.*


/**
 * Created by Aeolus on 2018/8/24.
 */

class HourWeatherView(context: Context, attrs: AttributeSet) : View(context, attrs) {

    private val pointPaint = Paint()
    private val outPointPaint = Paint()
    private val textPaint = Paint()
    private val dividerPaint = Paint()
    private val tmpPaint = Paint()
    private val path = Path()
    private var mDst=Path()
    private var temPathMeasureSpec=PathMeasure()





    private var tmpList: MutableList<Int> = ArrayList()
    private var popList: MutableList<Int> = ArrayList()
    private var dateList: MutableList<String> = ArrayList()
    private var txtList: MutableList<String> = ArrayList()
    private var dirList: MutableList<String> = ArrayList()

    private var viewHigh = 0 //控件高度
    private var viewWidth = 0 //控件宽度
    private var xUnit = 0f //x轴单位长度
    private var yUnit = 0f //y轴单位长度
    private var xPart = 8f //x轴等分
    private var textHigh = 0f //文字高度
    private var textSpace = 0f //文字间隔
    private var max= 0
    private var min = 0
    private var curvature=0.13f // 0<curvature<0.5
    private var mLength= 0f
    private var mAnimatorValue= 0f

    init {
        init()
    }

    private fun init() {

        pointPaint.color = ContextCompat.getColor(context, R.color.primary_black_text)
        pointPaint.isAntiAlias = true
        pointPaint.strokeWidth = pointWidth
        pointPaint.style = Paint.Style.FILL


        outPointPaint.color = ContextCompat.getColor(context, R.color.primary_black_text)
        outPointPaint.isAntiAlias = true
        outPointPaint.strokeWidth = outPointWidth
        outPointPaint.style = Paint.Style.STROKE

        tmpPaint.color = ContextCompat.getColor(context, R.color.primary_black_text)
        tmpPaint.isAntiAlias = true
        tmpPaint.strokeWidth = lineWidth
        tmpPaint.style = Paint.Style.STROKE

        textPaint.color = ContextCompat.getColor(context, R.color.primary_black_text)
        textPaint.isAntiAlias = true
        textPaint.textAlign = Paint.Align.CENTER

        dividerPaint.color = ContextCompat.getColor(context, R.color.hint_black_text)
        dividerPaint.isAntiAlias = true
        dividerPaint.strokeMiter= dividerWidth
        dividerPaint.textAlign = Paint.Align.CENTER

        val valueAnimator = ValueAnimator.ofFloat(0f, 1f)
        valueAnimator.addUpdateListener { valueAnimator ->
            mAnimatorValue = valueAnimator.animatedValue as Float
            invalidate()
        }
        valueAnimator.duration = 3000
        valueAnimator.start()
    }

    override fun onDraw(canvas: Canvas) {
        viewHigh = measuredHeight
        viewWidth = measuredWidth
        textPaint.textSize = viewWidth / (4*xPart)//按每格4个字计算文字大小
        val fm = textPaint.fontMetrics
        textHigh = fm.bottom - fm.top
        textSpace=0.5f*textHigh
        xUnit = viewWidth / (2*xPart)
        yUnit = getY(5,7) / 50
        if (tmpList.isNotEmpty()) {
            max = Collections.max(tmpList)
            min = Collections.min(tmpList)
            yUnit = getY(5,7) / (max - min)
        }

        //实时温度曲线
        drawTmpCurve(canvas)

        //日期
        for (i in dateList.indices) {
            val data1 = dateList[i].split(" ".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
            canvas.drawText(data1[1],  getBezierX(i), textHigh, textPaint)
        }

        //天气状况
        for (i in txtList.indices) {
            canvas.drawText(txtList[i], getBezierX(i), getY(2,4), textPaint)
        }

        //天气分割线
        for (i in 1 until dateList.size) {
            canvas.drawLine(getBezierX(i)-xUnit, textHigh, getBezierX(i)-xUnit, getY(3,4), dividerPaint)
        }

        //降雨分割线
        canvas.drawLine(0f,getY(2,3) , viewWidth.toFloat(), getY(2,3), dividerPaint)
        canvas.drawText("降水概率", viewWidth/2f, getY(1,2), textPaint)

        //降水概率
        for (i in popList.indices) {
            canvas.drawText(popList[i].toString() + "%", getBezierX(i ),getY(0,1), textPaint)
        }
    }

    private fun drawTmpCurve(canvas: Canvas){
        for (i in tmpList.indices) {

            var x1 = 0f
            var y1 = 0f

            var x2 = 0f
            var y2 = 0f

            when{
                i == 0 -> {
                    x1=getBezierX(i)+(getBezierX(i + 1)-getBezierX(i))*curvature
                    y1=getBezierY(i,tmpList)+(getBezierY(i + 1,tmpList)-getBezierY(i,tmpList))*curvature

                    x2=getBezierX(i+1)-(getBezierX(i + 2)-getBezierX(i))*curvature
                    y2=getBezierY(i+1,tmpList)-(getBezierY(i + 2,tmpList)-getBezierY(i,tmpList))*curvature

                    path.moveTo(getBezierX(i), getBezierY(i, tmpList))
                    path.cubicTo(x1, y1, x2, y2, getBezierX(i + 1), getBezierY(i + 1, tmpList))
                }
                i< tmpList.size - 2 -> {
                    x1=getBezierX(i)+(getBezierX(i + 1)-getBezierX(i-1))*curvature
                    y1=getBezierY(i,tmpList)+(getBezierY(i + 1,tmpList)-getBezierY(i-1,tmpList))*curvature

                    x2=getBezierX(i+1)-(getBezierX(i + 2)-getBezierX(i))*curvature
                    y2=getBezierY(i+1,tmpList)-(getBezierY(i + 2,tmpList)-getBezierY(i,tmpList))*curvature

                    path.cubicTo(x1, y1, x2, y2, getBezierX(i + 1), getBezierY(i + 1, tmpList))

                }
                i == tmpList.size - 2 -> {
                    x1=getBezierX(i)+(getBezierX(i + 1)-getBezierX(i-1))*curvature
                    y1=getBezierY(i,tmpList)+(getBezierY(i + 1,tmpList)-getBezierY(i-1,tmpList))*curvature

                    x2=getBezierX(i+1)-(getBezierX(i + 1)-getBezierX(i))*curvature
                    y2=getBezierY(i+1,tmpList)-(getBezierY(i + 1,tmpList)-getBezierY(i,tmpList))*curvature

                    path.cubicTo(x1, y1, x2, y2, getBezierX(i + 1), getBezierY(i + 1, tmpList))

                }
            }
            canvas.drawText(tmpList[i].toString() + "°", getBezierX(i), getBezierY(i, tmpList)-textSpace, textPaint)
        }
        temPathMeasureSpec.setPath(path, false)
        mLength = temPathMeasureSpec.length
        mDst.reset()
        mDst.lineTo(0f,0f)
        val stop = mLength * mAnimatorValue
        temPathMeasureSpec.getSegment(0f,stop,mDst,true)
        canvas.drawPath(mDst, tmpPaint)

    }


    private fun drawTmpBrokenLine(canvas: Canvas){
        for (i in tmpList.indices) {
            when {
                i == 0 -> {
                    path.moveTo(getBezierX(2*i+1), getBezierY(i,tmpList))
                    path.lineTo(getBezierX(2*i+3), getBezierY(i,tmpList))
                }
                i<tmpList.size-1 -> {
                    path.lineTo(getBezierX(2*i+3), getBezierY(i,tmpList))
                }
            }
            //            canvas.drawCircle(getBezierX(2*i+1),getBezierY(i,tmpList),pointRadius,pointPaint);
            //            canvas.drawCircle(getBezierX(2*i+1),getBezierY(i,tmpList),outPointRadius,outPointPaint);
            canvas.drawText(tmpList[i].toString() + "°", getBezierX(2 * i + 1), getBezierY(i, tmpList) - 20, textPaint)
        }
        temPathMeasureSpec.setPath(path, false)
        mLength = temPathMeasureSpec.length
        mDst.reset()
        mDst.lineTo(0f,0f)
        val stop = mLength * mAnimatorValue
        temPathMeasureSpec.getSegment(0f,stop,mDst,true)
        canvas.drawPath(mDst, tmpPaint)
    }

    fun loadViewData(hourlyList: List<Weather.HourlyWeather>?) {
        if (hourlyList != null) {
            tmpList.clear()
            popList.clear()
            dateList.clear()
            txtList.clear()
            dirList.clear()
            for (hourly in hourlyList){
                (tmpList as ArrayList).add(hourly.tmp.toInt())
                (popList as ArrayList).add(hourly.pop.toInt())
                hourly.time.let { (dateList as ArrayList).add(it) }
                hourly.cond_txt.let { (txtList as ArrayList).add(it) }
                hourly.wind_dir.let { (dirList as ArrayList).add(it) }
            }
        }
        invalidate()
    }

    private fun getBezierX(i: Int): Float {
        return (2*i+1) * xUnit
    }

    private fun getBezierY(i: Int, list: List<Int>): Float {
        return (max - list[i]) * yUnit + 2 * textHigh+2*textSpace
    }

    private fun getY(i1: Int,i2:Int): Float {
        return viewHigh - (i1 * textHigh+i2*textSpace)
    }


    companion object {

        private const val dividerWidth = 1f
        private const val lineWidth = 4f
        private const val pointWidth = 3f
        private const val outPointWidth = 3f
        private const val pointRadius = 5f
        private const val outPointRadius = 10f
    }

}

